File: qucanft/houses.py
"""
Astrological houses calculation module.

This module provides functionality to calculate astrological houses using
different house systems (Placidus, Koch, Equal, etc.) and determine
planetary house positions.
"""

from typing import Dict, List, Optional, Tuple
import numpy as np
import pandas as pd
from datetime import datetime
import math


class HousesCalculator:
    """
    A class for calculating astrological houses and planetary house positions.
    
    This class supports multiple house systems and provides methods to
    calculate house cusps and determine which house planets are in.
    """
    
    # House systems available
    HOUSE_SYSTEMS = {
        'placidus': 'Placidus',
        'koch': 'Koch',
        'equal': 'Equal House',
        'whole': 'Whole Sign',
        'campanus': 'Campanus',
        'regiomontanus': 'Regiomontanus'
    }
    
    # House meanings and themes
    HOUSE_MEANINGS = {
        1: {'name': 'Ascendant/Self', 'theme': 'Identity, appearance, first impressions'},
        2: {'name': 'Resources', 'theme': 'Money, possessions, values, self-worth'},
        3: {'name': 'Communication', 'theme': 'Communication, siblings, short trips, learning'},
        4: {'name': 'Home/Family', 'theme': 'Home, family, roots, emotional foundation'},
        5: {'name': 'Creativity', 'theme': 'Creativity, romance, children, self-expression'},
        6: {'name': 'Work/Health', 'theme': 'Work, health, daily routines, service'},
        7: {'name': 'Partnerships', 'theme': 'Marriage, partnerships, open enemies'},
        8: {'name': 'Transformation', 'theme': 'Death, rebirth, shared resources, occult'},
        9: {'name': 'Philosophy', 'theme': 'Higher learning, philosophy, long journeys'},
        10: {'name': 'Career/Status', 'theme': 'Career, reputation, public standing'},
        11: {'name': 'Friends/Goals', 'theme': 'Friends, groups, hopes, wishes'},
        12: {'name': 'Subconscious', 'theme': 'Subconscious, karma, hidden enemies, sacrifice'}
    }
    
    def __init__(self):
        """Initialize the HousesCalculator."""
        pass
    
    def calculate_ascendant(self, 
                          local_sidereal_time: float,
                          latitude: float,
                          obliquity: float = 23.4367) -> float:
        """
        Calculate the Ascendant (1st house cusp) for given time and location.
        
        Args:
            local_sidereal_time: Local sidereal time in hours
            latitude: Geographic latitude in degrees
            obliquity: Obliquity of the ecliptic in degrees
            
        Returns:
            Ascendant position in degrees of ecliptic longitude
        """
        # Convert to radians
        lst_rad = math.radians(local_sidereal_time * 15)  # LST in degrees
        lat_rad = math.radians(latitude)
        obl_rad = math.radians(obliquity)
        
        # Calculate ascendant using spherical trigonometry
        # This is a simplified calculation - in practice, you'd use more precise algorithms
        y = -math.cos(lst_rad)
        x = math.sin(lst_rad) * math.cos(obl_rad) + math.tan(lat_rad) * math.sin(obl_rad)
        
        ascendant_rad = math.atan2(y, x)
        ascendant_deg = math.degrees(ascendant_rad)
        
        # Normalize to 0-360 range
        return ascendant_deg % 360
    
    def calculate_midheaven(self, 
                           local_sidereal_time: float,
                           obliquity: float = 23.4367) -> float:
        """
        Calculate the Midheaven (10th house cusp) for given time.
        
        Args:
            local_sidereal_time: Local sidereal time in hours
            obliquity: Obliquity of the ecliptic in degrees
            
        Returns:
            Midheaven position in degrees of ecliptic longitude
        """
        # Convert LST to degrees
        lst_deg = local_sidereal_time * 15
        
        # Calculate MC using spherical trigonometry
        # This is a simplified calculation
        mc_deg = lst_deg % 360
        
        return mc_deg
    
    def calculate_equal_houses(self, ascendant: float) -> Dict[int, float]:
        """
        Calculate house cusps using the Equal House system.
        
        Args:
            ascendant: Ascendant position in degrees
            
        Returns:
            Dictionary with house numbers as keys and cusp positions as values
        """
        houses = {}
        
        for house_num in range(1, 13):
            # In Equal House system, each house is exactly 30 degrees
            cusp = (ascendant + (house_num - 1) * 30) % 360
            houses[house_num] = cusp
        
        return houses
    
    def calculate_whole_sign_houses(self, ascendant: float) -> Dict[int, float]:
        """
        Calculate house cusps using the Whole Sign system.
        
        Args:
            ascendant: Ascendant position in degrees
            
        Returns:
            Dictionary with house numbers as keys and cusp positions as values
        """
        # In Whole Sign system, the 1st house starts at 0° of the ascendant sign
        ascendant_sign_start = (int(ascendant // 30)) * 30
        
        houses = {}
        for house_num in range(1, 13):
            cusp = (ascendant_sign_start + (house_num - 1) * 30) % 360
            houses[house_num] = cusp
        
        return houses
    
    def calculate_placidus_houses(self, 
                                 ascendant: float,
                                 midheaven: float,
                                 latitude: float,
                                 obliquity: float = 23.4367) -> Dict[int, float]:
        """
        Calculate house cusps using the Placidus system (simplified).
        
        Args:
            ascendant: Ascendant position in degrees
            midheaven: Midheaven position in degrees
            latitude: Geographic latitude in degrees
            obliquity: Obliquity of the ecliptic in degrees
            
        Returns:
            Dictionary with house numbers as keys and cusp positions as values
            
        Note:
            This is a simplified implementation. A full Placidus calculation
            requires complex iterative algorithms and is typically done using
            specialized libraries like Swiss Ephemeris.
        """
        houses = {}
        
        # Set the angular houses (1st, 4th, 7th, 10th)
        houses[1] = ascendant
        houses[4] = (midheaven + 180) % 360  # IC (Imum Coeli)
        houses[7] = (ascendant + 180) % 360  # Descendant
        houses[10] = midheaven  # MC (Medium Coeli)
        
        # Simplified calculation for intermediate houses
        # In practice, this would use complex trigonometry
        for house_num in [2, 3, 5, 6, 8, 9, 11, 12]:
            if house_num in [2, 3]:
                # Houses 2 and 3 between ASC and MC
                factor = (house_num - 1) / 3
                angle_diff = (midheaven - ascendant) % 360
                if angle_diff > 180:
                    angle_diff = 360 - angle_diff
                houses[house_num] = (ascendant + factor * angle_diff) % 360
            elif house_num in [5, 6]:
                # Houses 5 and 6 between MC and DSC
                factor = (house_num - 4) / 3
                angle_diff = (houses[7] - midheaven) % 360
                if angle_diff > 180:
                    angle_diff = 360 - angle_diff
                houses[house_num] = (midheaven + factor * angle_diff) % 360
            elif house_num in [8, 9]:
                # Houses 8 and 9 between DSC and IC
                factor = (house_num - 7) / 3
                angle_diff = (houses[4] - houses[7]) % 360
                if angle_diff > 180:
                    angle_diff = 360 - angle_diff
                houses[house_num] = (houses[7] + factor * angle_diff) % 360
            elif house_num in [11, 12]:
                # Houses 11 and 12 between IC and ASC
                factor = (house_num - 10) / 3
                angle_diff = (ascendant - houses[4]) % 360
                if angle_diff > 180:
                    angle_diff = 360 - angle_diff
                houses[house_num] = (houses[4] + factor * angle_diff) % 360
        
        return houses
    
    def calculate_houses(self, 
                        ascendant: float,
                        midheaven: Optional[float] = None,
                        latitude: Optional[float] = None,
                        house_system: str = 'equal') -> Dict[int, float]:
        """
        Calculate house cusps using the specified house system.
        
        Args:
            ascendant: Ascendant position in degrees
            midheaven: Midheaven position in degrees (required for some systems)
            latitude: Geographic latitude in degrees (required for some systems)
            house_system: House system to use ('equal', 'whole', 'placidus', etc.)
            
        Returns:
            Dictionary with house numbers as keys and cusp positions as values
        """
        house_system = house_system.lower()
        
        if house_system == 'equal':
            return self.calculate_equal_houses(ascendant)
        elif house_system == 'whole':
            return self.calculate_whole_sign_houses(ascendant)
        elif house_system == 'placidus':
            if midheaven is None or latitude is None:
                raise ValueError("Placidus system requires midheaven and latitude")
            return self.calculate_placidus_houses(ascendant, midheaven, latitude)
        else:
            # Default to equal house system
            print(f"Warning: House system '{house_system}' not implemented, using Equal House")
            return self.calculate_equal_houses(ascendant)
    
    def determine_planet_house(self, 
                             planet_longitude: float,
                             house_cusps: Dict[int, float]) -> int:
        """
        Determine which house a planet is in based on its longitude.
        
        Args:
            planet_longitude: Planet's ecliptic longitude in degrees
            house_cusps: Dictionary of house cusps
            
        Returns:
            House number (1-12) that the planet is in
        """
        # Normalize planet longitude to 0-360
        planet_longitude = planet_longitude % 360
        
        for house_num in range(1, 13):
            cusp_current = house_cusps[house_num]
            cusp_next = house_cusps[house_num + 1] if house_num < 12 else house_cusps[1]
            
            # Handle the case where the house crosses 0°
            if cusp_current > cusp_next:
                # House crosses 0°
                if planet_longitude >= cusp_current or planet_longitude < cusp_next:
                    return house_num
            else:
                # Normal case
                if cusp_current <= planet_longitude < cusp_next:
                    return house_num
        
        # Fallback (shouldn't happen with correct calculations)
        return 1
    
    def add_house_positions(self, 
                           planetary_data: pd.DataFrame,
                           house_cusps: Dict[int, float]) -> pd.DataFrame:
        """
        Add house positions to planetary data.
        
        Args:
            planetary_data: DataFrame with planetary positions
            house_cusps: Dictionary of house cusps
            
        Returns:
            DataFrame with added house position columns
        """
        if 'Ecliptic_Longitude' not in planetary_data.columns:
            raise ValueError("DataFrame must contain 'Ecliptic_Longitude' column")
        
        result_df = planetary_data.copy()
        
        # Calculate house positions
        house_positions = []
        house_meanings = []
        
        for _, row in planetary_data.iterrows():
            house_num = self.determine_planet_house(row['Ecliptic_Longitude'], house_cusps)
            house_positions.append(house_num)
            house_meanings.append(self.HOUSE_MEANINGS[house_num]['theme'])
        
        # Add house columns
        result_df['House'] = house_positions
        result_df['House_Meaning'] = house_meanings
        
        return result_df
    
    def get_house_info(self, house_number: int) -> Optional[Dict[str, str]]:
        """
        Get information about a specific house.
        
        Args:
            house_number: House number (1-12)
            
        Returns:
            Dictionary with house information, or None if invalid house number
        """
        if house_number in self.HOUSE_MEANINGS:
            return self.HOUSE_MEANINGS[house_number].copy()
        return None
    
    def calculate_house_strengths(self, 
                                planetary_data: pd.DataFrame,
                                house_cusps: Dict[int, float]) -> Dict[int, int]:
        """
        Calculate the number of planets in each house.
        
        Args:
            planetary_data: DataFrame with planetary positions
            house_cusps: Dictionary of house cusps
            
        Returns:
            Dictionary with house numbers as keys and planet counts as values
        """
        if 'Ecliptic_Longitude' not in planetary_data.columns:
            raise ValueError("DataFrame must contain 'Ecliptic_Longitude' column")
        
        house_counts = {i: 0 for i in range(1, 13)}
        
        for _, row in planetary_data.iterrows():
            house_num = self.determine_planet_house(row['Ecliptic_Longitude'], house_cusps)
            house_counts[house_num] += 1
        
        return house_counts
    
    def format_house_cusp(self, house_num: int, cusp_longitude: float) -> str:
        """
        Format a house cusp position as a readable string.
        
        Args:
            house_num: House number (1-12)
            cusp_longitude: Cusp longitude in degrees
            
        Returns:
            Formatted string with house number and position
        """
        # Convert longitude to zodiac position
        sign_index = int(cusp_longitude // 30)
        degree_in_sign = cusp_longitude % 30
        
        zodiac_signs = ['Aries', 'Taurus', 'Gemini', 'Cancer', 'Leo', 'Virgo',
                       'Libra', 'Scorpio', 'Sagittarius', 'Capricorn', 'Aquarius', 'Pisces']
        
        sign_name = zodiac_signs[sign_index]
        
        return f"House {house_num}: {int(degree_in_sign)}°{int((degree_in_sign % 1) * 60):02d}' {sign_name}"

File: pkg/houses/houses.go
// Package houses provides astrological house system calculations.
package houses

import (
	"fmt"
	"math"
	
	"github.com/Qucanft/Qucanft/pkg/coordinates"
	"github.com/Qucanft/Qucanft/pkg/planets"
	"github.com/Qucanft/Qucanft/pkg/zodiac"
)

// HouseSystem represents different house calculation systems
type HouseSystem string

const (
	// Equal houses - each house is exactly 30°
	Equal HouseSystem = "Equal"
	
	// Placidus - most common time-based system
	Placidus HouseSystem = "Placidus"
	
	// Whole Sign - each house occupies an entire zodiac sign
	WholeSign HouseSystem = "WholeSign"
	
	// Koch - alternative time-based system
	Koch HouseSystem = "Koch"
	
	// Campanus - space-based system
	Campanus HouseSystem = "Campanus"
	
	// Regiomontanus - medieval system
	Regiomontanus HouseSystem = "Regiomontanus"
)

// House represents an astrological house
type House struct {
	Number      int
	Name        string
	Theme       string
	Ruler       string
	CuspDegree  float64 // Ecliptic longitude of the house cusp
	Size        float64 // Size of the house in degrees
	Planets     []planets.Planet
	Description string
}

// HouseCalculator handles house calculations
type HouseCalculator struct {
	system HouseSystem
}

// NewHouseCalculator creates a new house calculator with the specified system
func NewHouseCalculator(system HouseSystem) *HouseCalculator {
	return &HouseCalculator{
		system: system,
	}
}

// SetSystem changes the house system
func (hc *HouseCalculator) SetSystem(system HouseSystem) {
	hc.system = system
}

// GetSystem returns the current house system
func (hc *HouseCalculator) GetSystem() HouseSystem {
	return hc.system
}

// CalculateHouseCusps calculates the house cusps for a given time and location
func (hc *HouseCalculator) CalculateHouseCusps(ascendant, midheaven, latitude float64) ([]float64, error) {
	switch hc.system {
	case Equal:
		return hc.calculateEqualHouses(ascendant), nil
	case Placidus:
		return hc.calculatePlacidusHouses(ascendant, midheaven, latitude), nil
	case WholeSign:
		return hc.calculateWholeSignHouses(ascendant), nil
	case Koch:
		return hc.calculateKochHouses(ascendant, midheaven, latitude), nil
	case Campanus:
		return hc.calculateCampanusHouses(ascendant, midheaven, latitude), nil
	case Regiomontanus:
		return hc.calculateRegiomontanusHouses(ascendant, midheaven, latitude), nil
	default:
		return nil, fmt.Errorf("unsupported house system: %s", hc.system)
	}
}

// CalculateHouses calculates complete house information
func (hc *HouseCalculator) CalculateHouses(ascendant, midheaven, latitude float64) ([]House, error) {
	cusps, err := hc.CalculateHouseCusps(ascendant, midheaven, latitude)
	if err != nil {
		return nil, err
	}
	
	houses := make([]House, 12)
	houseInfo := getHouseInformation()
	
	for i := 0; i < 12; i++ {
		nextCusp := cusps[(i+1)%12]
		size := nextCusp - cusps[i]
		if size < 0 {
			size += 360
		}
		
		houses[i] = House{
			Number:      i + 1,
			Name:        houseInfo[i].Name,
			Theme:       houseInfo[i].Theme,
			Ruler:       houseInfo[i].Ruler,
			CuspDegree:  cusps[i],
			Size:        size,
			Planets:     []planets.Planet{},
			Description: houseInfo[i].Description,
		}
	}
	
	return houses, nil
}

// AddPlanetsToHouses assigns planets to houses based on their positions
func (hc *HouseCalculator) AddPlanetsToHouses(houses []House, positions []planets.PlanetaryPosition) []House {
	// Create a copy of houses to avoid modifying the original
	result := make([]House, len(houses))
	copy(result, houses)
	
	// Clear existing planets
	for i := range result {
		result[i].Planets = []planets.Planet{}
	}
	
	// Assign planets to houses
	for _, pos := range positions {
		houseIndex := hc.findHouseForPosition(pos.Coordinates.Longitude, houses)
		if houseIndex >= 0 && houseIndex < len(result) {
			result[houseIndex].Planets = append(result[houseIndex].Planets, pos.Planet)
		}
	}
	
	return result
}

// findHouseForPosition finds which house a given ecliptic longitude belongs to
func (hc *HouseCalculator) findHouseForPosition(longitude float64, houses []House) int {
	longitude = coordinates.NormalizeAngle(longitude)
	
	for i, house := range houses {
		nextHouseIndex := (i + 1) % len(houses)
		nextHouseCusp := houses[nextHouseIndex].CuspDegree
		
		// Handle wrap-around at 360°/0°
		if house.CuspDegree <= nextHouseCusp {
			if longitude >= house.CuspDegree && longitude < nextHouseCusp {
				return i
			}
		} else {
			// House spans across 0° point
			if longitude >= house.CuspDegree || longitude < nextHouseCusp {
				return i
			}
		}
	}
	
	return -1 // Not found (shouldn't happen)
}

// calculateEqualHouses calculates equal house cusps
func (hc *HouseCalculator) calculateEqualHouses(ascendant float64) []float64 {
	cusps := make([]float64, 12)
	
	for i := 0; i < 12; i++ {
		cusps[i] = coordinates.NormalizeAngle(ascendant + float64(i)*30)
	}
	
	return cusps
}

// calculatePlacidusHouses calculates Placidus house cusps
func (hc *HouseCalculator) calculatePlacidusHouses(ascendant, midheaven, latitude float64) []float64 {
	cusps := make([]float64, 12)
	
	// Set the main angles
	cusps[0] = ascendant                                      // 1st house (Ascendant)
	cusps[3] = coordinates.NormalizeAngle(ascendant + 180)    // 4th house (IC)
	cusps[6] = coordinates.NormalizeAngle(ascendant + 180)    // 7th house (Descendant)
	cusps[9] = midheaven                                      // 10th house (MC)
	
	// Calculate intermediate houses using Placidus method
	latRad := latitude * coordinates.DegreesToRadians
	
	// Calculate 2nd and 3rd houses
	for i := 1; i <= 2; i++ {
		t := float64(i) / 3.0
		cusps[i] = hc.calculatePlacidusHouse(ascendant, midheaven, latRad, t)
	}
	
	// Calculate 5th and 6th houses
	for i := 4; i <= 5; i++ {
		t := float64(i-3) / 3.0
		cusps[i] = hc.calculatePlacidusHouse(cusps[3], cusps[6], latRad, t)
	}
	
	// Calculate 8th and 9th houses
	for i := 7; i <= 8; i++ {
		t := float64(i-6) / 3.0
		cusps[i] = hc.calculatePlacidusHouse(cusps[6], cusps[9], latRad, t)
	}
	
	// Calculate 11th and 12th houses
	for i := 10; i <= 11; i++ {
		t := float64(i-9) / 3.0
		cusps[i] = hc.calculatePlacidusHouse(cusps[9], cusps[0], latRad, t)
	}
	
	return cusps
}

// calculatePlacidusHouse calculates a single Placidus house cusp
func (hc *HouseCalculator) calculatePlacidusHouse(start, end, latitude, t float64) float64 {
	// This is a simplified Placidus calculation
	// Real implementation would involve more complex spherical trigonometry
	
	diff := end - start
	if diff < 0 {
		diff += 360
	}
	
	// Apply time-based adjustment
	adjustment := math.Sin(t*math.Pi/2) * math.Tan(latitude) * 5 // Simplified
	
	result := start + diff*t + adjustment
	return coordinates.NormalizeAngle(result)
}

// calculateWholeSignHouses calculates Whole Sign house cusps
func (hc *HouseCalculator) calculateWholeSignHouses(ascendant float64) []float64 {
	cusps := make([]float64, 12)
	
	// Find the zodiac sign of the ascendant
	zc := zodiac.NewZodiacCalculator()
	ascendantZodiac := zc.EclipticToZodiac(ascendant)
	
	// Each house starts at the beginning of a zodiac sign
	startSign := int(ascendantZodiac.Sign.StartDeg / 30)
	
	for i := 0; i < 12; i++ {
		signIndex := (startSign + i) % 12
		cusps[i] = float64(signIndex * 30)
	}
	
	return cusps
}

// calculateKochHouses calculates Koch house cusps
func (hc *HouseCalculator) calculateKochHouses(ascendant, midheaven, latitude float64) []float64 {
	// Koch system is similar to Placidus but with different calculation method
	// This is a simplified implementation
	return hc.calculatePlacidusHouses(ascendant, midheaven, latitude)
}

// calculateCampanusHouses calculates Campanus house cusps
func (hc *HouseCalculator) calculateCampanusHouses(ascendant, midheaven, latitude float64) []float64 {
	// Campanus is a space-based system
	// This is a simplified implementation
	cusps := make([]float64, 12)
	
	// Use equal division as a base and apply spatial adjustments
	base := hc.calculateEqualHouses(ascendant)
	
	for i := 0; i < 12; i++ {
		// Apply spatial adjustment based on latitude
		adjustment := math.Sin(float64(i)*math.Pi/6) * latitude * 0.1
		cusps[i] = coordinates.NormalizeAngle(base[i] + adjustment)
	}
	
	return cusps
}

// calculateRegiomontanusHouses calculates Regiomontanus house cusps
func (hc *HouseCalculator) calculateRegiomontanusHouses(ascendant, midheaven, latitude float64) []float64 {
	// Regiomontanus is a medieval system
	// This is a simplified implementation
	return hc.calculatePlacidusHouses(ascendant, midheaven, latitude)
}

// GetHousePosition returns the house position for a given ecliptic longitude
func (hc *HouseCalculator) GetHousePosition(longitude float64, houses []House) (int, float64, error) {
	houseIndex := hc.findHouseForPosition(longitude, houses)
	if houseIndex < 0 {
		return -1, 0, fmt.Errorf("could not determine house for longitude %.2f", longitude)
	}
	
	// Calculate position within the house (0-1)
	house := houses[houseIndex]
	nextHouseIndex := (houseIndex + 1) % len(houses)
	nextHouseCusp := houses[nextHouseIndex].CuspDegree
	
	var positionInHouse float64
	if house.CuspDegree <= nextHouseCusp {
		positionInHouse = (longitude - house.CuspDegree) / (nextHouseCusp - house.CuspDegree)
	} else {
		// Handle wrap-around
		if longitude >= house.CuspDegree {
			positionInHouse = (longitude - house.CuspDegree) / (360 - house.CuspDegree + nextHouseCusp)
		} else {
			positionInHouse = (360 - house.CuspDegree + longitude) / (360 - house.CuspDegree + nextHouseCusp)
		}
	}
	
	return houseIndex + 1, positionInHouse, nil
}

// HouseInfo represents basic information about a house
type HouseInfo struct {
	Name        string
	Theme       string
	Ruler       string
	Description string
}

// getHouseInformation returns basic information about the 12 houses
func getHouseInformation() []HouseInfo {
	return []HouseInfo{
		{
			Name:        "1st House",
			Theme:       "Self, Identity, Appearance",
			Ruler:       "Mars",
			Description: "The house of self, personality, and how you appear to others",
		},
		{
			Name:        "2nd House", 
			Theme:       "Money, Possessions, Values",
			Ruler:       "Venus",
			Description: "The house of personal resources, money, and material possessions",
		},
		{
			Name:        "3rd House",
			Theme:       "Communication, Siblings, Short Trips",
			Ruler:       "Mercury",
			Description: "The house of communication, learning, and immediate environment",
		},
		{
			Name:        "4th House",
			Theme:       "Home, Family, Roots",
			Ruler:       "Moon",
			Description: "The house of home, family, and emotional foundation",
		},
		{
			Name:        "5th House",
			Theme:       "Creativity, Romance, Children",
			Ruler:       "Sun",
			Description: "The house of creativity, romance, and self-expression",
		},
		{
			Name:        "6th House",
			Theme:       "Work, Health, Daily Routine",
			Ruler:       "Mercury",
			Description: "The house of work, health, and daily responsibilities",
		},
		{
			Name:        "7th House",
			Theme:       "Partnerships, Marriage, Others",
			Ruler:       "Venus",
			Description: "The house of partnerships, marriage, and open enemies",
		},
		{
			Name:        "8th House",
			Theme:       "Transformation, Shared Resources, Death",
			Ruler:       "Mars",
			Description: "The house of transformation, shared resources, and hidden things",
		},
		{
			Name:        "9th House",
			Theme:       "Philosophy, Higher Learning, Travel",
			Ruler:       "Jupiter",
			Description: "The house of higher learning, philosophy, and long-distance travel",
		},
		{
			Name:        "10th House",
			Theme:       "Career, Reputation, Authority",
			Ruler:       "Saturn",
			Description: "The house of career, reputation, and public standing",
		},
		{
			Name:        "11th House",
			Theme:       "Friends, Groups, Hopes",
			Ruler:       "Uranus",
			Description: "The house of friends, groups, and hopes and wishes",
		},
		{
			Name:        "12th House",
			Theme:       "Spirituality, Subconscious, Hidden",
			Ruler:       "Neptune",
			Description: "The house of spirituality, subconscious, and hidden enemies",
		},
	}
}

// String methods
func (h House) String() string {
	planetNames := make([]string, len(h.Planets))
	for i, planet := range h.Planets {
		planetNames[i] = planet.Name
	}
	return fmt.Sprintf("%s (%.1f°): %s - Planets: %v", h.Name, h.CuspDegree, h.Theme, planetNames)
}

func (hs HouseSystem) String() string {
	return string(hs)
}